### ISSUE TITLE
Upgrade RateLimiterRedis to use node-redis v4 as storage client

### ISSUE THREAD

**Issue by 0x1C1B** [animir/node-rate-limiter-flexible#155]
More or less two months ago, a new version of [node-redis](https://www.npmjs.com/package/redis) was released. This makes version `4.x.x` the current stable version. For compatibility reasons, the [RateLimiterRedis](https://github.com/animir/node-rate-limiter-flexible/wiki/Redis) implementation should be updated to the latest version. The new [node-redis](https://www.npmjs.com/package/redis) version works with promises by default and makes promisify superfluous.

The new version of [node-redis](https://www.npmjs.com/package/redis) also brought changes to the API. Hence the question, is backwards compatibility required for version `3.x.x` or could the code be easily adapted for [node-redis](https://www.npmjs.com/package/redis) `4.x.x` in the next [node-rate-limiter-flexible](https://github.com/animir/node-rate-limiter-flexible) version? Anyway, the update should be completed, since many projects that already rely on redis for caching (e.g. database query caching) will use [node-redis](https://www.npmjs.com/package/redis) `4.x.x` in the future. Otherwise, these projects would be forced to use [node-redis](https://www.npmjs.com/package/redis) `3.x.x` because of the rate limiting. Any opinions on this?

**Comment by animir** [comment]
@0x1C1B Hey, thank you for rising this topic.
First of all, we should understand, how node-redis v4 release changes affect current approach implemented in RateLimiterRedis and how much efforts will be required to refactor it.
If there is too much refactoring, we can prepare `rate-limiter-flexible` v3 major update and deprecate node-redis v3 in favor of node-redis v4.
It would be good if we can fix RateLimiterRedis to support both v3 and v4.

Note, RateLimiterRedis code also supports `ioredis`, so if we go with refactoring all packages:`ioredis`, `node-redis` v3 and `node-redis` v4 should be supported.

**Comment by 0x1C1B** [comment]
Sure, [node-redis](https://github.com/redis/node-redis) published a [migration guide](https://github.com/redis/node-redis/blob/master/docs/v3-to-v4.md). Version `4.x.x` comes with some breaking changes, some affecting this project others not.

- **Client Configuration:** The API for [client configuration](https://github.com/redis/node-redis/blob/master/docs/client-configuration.md) using `createClient` changed. In my opinion, this shouldn't affect this project, however, since the [constructor of RateLimiterRedis.js](https://github.com/animir/node-rate-limiter-flexible/blob/master/lib/RateLimiterRedis.js#L24) accepts a client instance and does not create it by its own. Right?
- **Auto-Connect:** The auto-connect functionality when creating the client instance is omitted. Instead, the connection must now be established manually via `connect()`. Should also not affect the project for the above reason.
- **Events:** Many events are also omitted in version `4.x.x`. A list can be found [here](https://github.com/redis/node-redis/blob/master/docs/v3-to-v4.md#all-the-removed-events). However, as far as I can see from the code, this implementation doesn't use events, right?
- **Promises:** The big change affecting this project is Promises. In the new version, [node-redis](https://github.com/redis/node-redis) uses promises by default and no longer supports callbacks.

To sum it up, the sticking point will be the promises. The new API of node-redis `4.x.x` no longer supports callbacks while node-redis `3.x.x` only knows callbacks. @animir Because you mentioned ioredis, I checked and ioredis supports both callbacks and promises. Means with the functionality that this project uses, I have not found any points in which ioredis and node-redis `4.x.x` would differ significantly. Switching to promises would only be a problem for node-redis `3.x.x`. There is the question of dropping support in the next major release or working with promisify as is already being done.

**Comment by animir** [comment]
@0x1C1B Hey, thank you for the detailed analysis. I think, you're right about promises and everything. Have you tried `legacyMode: true` for `node-redis`? It is mentioned in [their migration guide](https://github.com/redis/node-redis/blob/master/docs/v3-to-v4.md#legacy-mode). May be it is fine to live with it a while, until `RateLimiterRedis` is completely refactored for promises in the next v3 major `rate-limiter-flexible` release.
